test/autograd/test_fallback.py

```diff
@@ -241,6 +242,131 @@ def backward(ctx, gx):
                 gx, = torch.autograd.grad(z, x, torch.ones_like(z))
                 self.assertEqual(gx, expected)
 
+    @parametrize("mode", ("nothing", "warn"))
+    def test_inplace_on_tensor_that_does_not_require_grad(self, mode):
+        # We don't do anything special (that is, we don't rebase history).
+        # See NOTE [autograd fallback and in-place operations] for why
+        with autograd_fallback_mode(mode):
+            lib = Library(self.test_ns, "FRAGMENT")
+
+            # Correct usage of (a!)
+            lib.define("foo(Tensor(a!) self, Tensor other) -> Tensor(a!)")
+
+            def foo_impl(x, y):
+                x_d = x.detach()
+                y = y.detach()
+                x_d.add_(y)
+                return x
+
+            lib.impl("foo", foo_impl, "CPU")
+            op = self.get_op("foo")
+
+            x = torch.randn(3)
+            y = torch.randn(3, requires_grad=True)
+            with self.assertRaisesRegex(RuntimeError, "does not require grad"):
+                z = x.clone()
+                op(z, y)
+                torch.autograd.grad(z, y, torch.ones_like(z), allow_unused=True)
```

Comment:
nit:
- we could probably deduplicate everything except the function definition since they are all called the same way
- we could also move everything except the grad/backward call from under assertRaises 

(same below)

Reply:
We could also combine the two tests if you'd like